### ISSUE TITLE
Rclone's bash_completion file seems to break general autocompletion and is way too big

### ISSUE THREAD

**Issue by DanielGibson** [rclone/rclone#7000]
#### What is the problem you are having with rclone?

The bash_completion file is 4.2MB, which is about two orders of magnitude above what would be acceptable (for reference, the largest other ones on my system are of git with 77KB and ipset with 67KB) and for that reason takes pretty long to load.

Furthermore, while it seems to work if I source it manually directly after generation (`. /etc/bash_completion.d/rclone`), after a fresh log in, bash completion is broken for *all* commands (only completes to filenames in current dir), until I remove /etc/bash_completion.d/rclone and log in again.

#### What is your rclone version (output from `rclone version`)

rclone v1.62.2
- os/version: ubuntu 22.04 (64 bit)
- os/kernel: 5.15.0-71-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.20.2
- go/linking: static
- go/tags: none

installed from https://github.com/rclone/rclone/releases/download/v1.62.2/rclone-v1.62.2-linux-amd64.deb

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)

Ubuntu 22.04 amd64

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone genautocomplete bash`

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

```
# rclone -vv genautocomplete bash
2023/05/11 14:37:54 DEBUG : rclone: Version "v1.62.2" starting with parameters ["rclone" "-vv" "genautocomplete" "bash"]
2023/05/11 14:37:54 DEBUG : rclone: Version "v1.62.2" finishing with parameters ["rclone" "-vv" "genautocomplete" "bash"]
```

#### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

**Comment by animosity22** [comment]
No issues at all when I use it.

It's big because there are a lot of commands :)

```
[felix@gemini bash_completion.d]$ du -sh *
8.0K	authselect-completion.sh
4.3M	rclone
```

Can you be more specific on what error you are seeing and how to reproduce it?

**Comment by DanielGibson** [comment]
> It's big because there are a lot of commands :)

git also has a lot of commands and still is only 1/55th the size.

4.2 MB is **a lot** of text, about the size of the bible (see https://www.gutenberg.org/ebooks/10). I'm pretty sure it doesn't have to be this big..

The problem I'm seeing (apart from login taking longer because that file needs to be read and parsed first) is that normal bash completion doesn't work anymore after logging out and logging in again, neither for rclone nor for other commands.

However, for some reason I currently can't reproduce the issue anymore, it now works.
No idea why, could it depend on the order the scripts in /etc/bash_completion.d/ (and whatever else is used) are included? What does the order depend on, is it alphabetical or (likely?) "whatever is first in the filesystem"?

**Comment by ncw** [comment]
We use the bash completion auto generator from the cobra package: https://github.com/spf13/cobra/blob/main/shell_completions.md

One of the problems is that rclone has a lot of global flags

```
$ rclone help flags . | wc -l
750
```

And these are written out in the file for each rclone command.

This makes the completion file very repetitive.

```
$ wc -l /etc/bash_completion.d/rclone 
107105 /etc/bash_completion.d/rclone
$ sort </etc/bash_completion.d/rclone | uniq | wc -l
2336
```

If someone wants to fix this, it needs to be fixed in https://github.com/spf13/cobra/

**Comment by memeplex** [comment]
I've reported this to cobra's tracker.

**Comment by marckhouzam** [comment]
Please use GenBashCompletionV2() instead of the legacy GenBashCompletion()

**Comment by memeplex** [comment]
Thanks for the tip @marckhouzam ? Is this V2 variant related to bash_completion 2 or are they unrelated? Will it work with legacy bash_completion?

**Comment by marckhouzam** [comment]
They are independent. The cobra bash v2 solution no longer generates the exhaustive list of completions. Instead, it calls into the actual program to get the completions.

it does have one limitation: if you use legacy custom completions (the ones written in bash), those won’t work. You will need to rewrite them in go. Please see https://github.com/spf13/cobra/blob/main/site/content/completions/_index.md#bash-completion-v2

**Comment by ncw** [comment]
I did have an experiment with the new cobra completion to get it to work in other shells. I'm not sure whether that is exactly the v2 completion you are talking about but it calls rclone to make the completions which seems much more sensible to me.

[v1.66.0-beta.7686.d49391db6.fix-completion](https://beta.rclone.org/branch/fix-completion/v1.66.0-beta.7686.d49391db6.fix-completion/) on branch [fix-completion](https://github.com/rclone/rclone/tree/fix-completion) (uploaded in 15-30 mins)

Is this running along the right lines ? Anyone want to have a go with it? Here is the commit https://github.com/rclone/rclone/commit/d49391db6232f61de737d15c8d3b8639dfb1e060

**Comment by marckhouzam** [comment]
Nice @ncw. I can’t speak to the validity of the change but from a cobra perpective that is indeed what you want to do. I’ve this is in you can start using `GenBashCompletionV2()` pretty much transparently.

**Comment by ncw** [comment]
I rebased the completions patch and updated it to use `GenBashCompletionV2` and this seems to fix the problems I was having with it so I think this is ready to go.

The completions file is tiny now and all the hard work is being done in rclone rather than in bash which is a lot easier for me to reason about!

Any testing much appreciated :-)

[v1.66.0-beta.7705.c53cc4e24.fix-completion](https://beta.rclone.org/branch/fix-completion/v1.66.0-beta.7705.c53cc4e24.fix-completion/) on branch [fix-completion](https://github.com/rclone/rclone/tree/fix-completion) (uploaded in 15-30 mins)

**Comment by ncw** [comment]
I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org/) in 15-30 minutes and released in v1.66
